File: java/org.financer.client.javafx/src/main/java/org/financer/client/javafx/main/FinancerApplication.java
package org.financer.client.javafx.main;

import javafx.application.Application;
import javafx.application.Platform;
import javafx.application.Preloader;
import javafx.beans.property.BooleanProperty;
import javafx.beans.property.SimpleBooleanProperty;
import javafx.fxml.FXMLLoader;
import javafx.scene.Parent;
import javafx.scene.Scene;
import javafx.scene.image.Image;
import javafx.scene.paint.Color;
import javafx.scene.text.Font;
import javafx.stage.Stage;
import javafx.stage.StageStyle;
import org.financer.client.domain.api.RestApi;
import org.financer.client.domain.api.RestApiImpl;
import org.financer.client.domain.model.category.CategoryRoot;
import org.financer.client.domain.model.user.User;
import org.financer.client.format.I18N;
import org.financer.client.javafx.local.LocalStorageImpl;
import org.financer.client.local.LocalStorage;
import org.financer.shared.domain.model.value.objects.SettingPair;

import java.io.IOException;
import java.io.Serializable;
import java.util.Locale;
import java.util.ResourceBundle;
import java.util.logging.Level;
import java.util.logging.Logger;

public class FinancerApplication extends Application {

    private final RestApi restApi = new RestApiImpl();
    private final LocalStorage localStorage = new LocalStorageImpl();
    private final Logger logger = Logger.getLogger("FinancerApplication");
    private BooleanProperty ready = new SimpleBooleanProperty(false);

    @Override
    public void init() {
        User user = LocalStorageImpl.getInstance().readObject("user");
        if (user != null) {
            notifyPreloader(new Preloader.ProgressNotification(0));
            restApi.getUsersCategories(user.getId(), categories ->
                    restApi.getUsersVariableTransactions(user.getId(), 0, variableTransactions ->
                            restApi.getUsersFixedTransactions(user.getId()).execute()).execute()).execute();
        }
    }

    @Override
    public void start(Stage primaryStage) throws IOException {
        this.logger.log(Level.INFO, "Financer Application has started.");

        User user = LocalStorageImpl.getInstance().readObject("user");
        if (user == null) {
            initLoginStage();
        } else {
            I18N.setLocalStorage(LocalStorageImpl.getInstance());

            ready.addListener((observableValue, oldValue, newValue) -> {
                if (Boolean.TRUE.equals(newValue)) {
                    Platform.runLater(() -> {
                        try {
                            initMainStage();
                        } catch (IOException e) {
                            logger.log(Level.SEVERE, e.getMessage(), e);
                        }
                    });
                }
            });
        }
    }

    private void initLoginStage() throws IOException {
        // setting up language
        ResourceBundle resourceBundle = ResourceBundle.getBundle("Financer", Locale.ENGLISH);

        // loading FXML file
        Parent root = FXMLLoader.load(getClass().getResource("/org/financer/client/javafx/main/views/login.fxml"), resourceBundle);
        Stage primaryStage = new Stage();
        Scene scene = new Scene(root, 800, 500);

        scene.getStylesheets().add(getClass().getResource("/org/financer/client/javafx/main/style/colors.style.css").toExternalForm());
        scene.getStylesheets().add(getClass().getResource("/org/financer/client/javafx/main/style/form.style.css").toExternalForm());
        scene.getStylesheets().add(getClass().getResource("/org/financer/client/javafx/main/style/login.style.css").toExternalForm());
        scene.setFill(Color.TRANSPARENT);

        primaryStage.getIcons().add(new Image(FinancerApplication.class.getResourceAsStream("/images/icons/financer-icon.png")));
        Font.loadFont(getClass().getResource("/fonts/Roboto-Regular.ttf").toExternalForm(), 12);
        Font.loadFont(getClass().getResource("/fonts/Roboto-Medium.ttf").toExternalForm(), 12);

        primaryStage.initStyle(StageStyle.TRANSPARENT);

        primaryStage.setTitle("Financer - Login");
        primaryStage.setScene(scene);
        primaryStage.show();
    }

    private void initMainStage() throws IOException {
        ResourceBundle resourceBundle = ResourceBundle.getBundle("Financer",
                (Locale) ((User) LocalStorageImpl.getInstance().readObject("user")).getValueOrDefault(SettingPair.Property.LANGUAGE));

        Parent root = FXMLLoader.load(getClass().getResource("views/financer.fxml"), resourceBundle);

        Stage primaryStage = new Stage();
        Scene scene = new Scene(root, 1300, 700);

        scene.getStylesheets().add(getClass().getResource("/org/financer/client/javafx/main/style/colors.style.css").toExternalForm());
        scene.getStylesheets().add(getClass().getResource("/org/financer/client/javafx/main/style/form.style.css").toExternalForm());
        scene.getStylesheets().add(getClass().getResource("/org/financer/client/javafx/main/style/navbar.style.css").toExternalForm());
        scene.getStylesheets().add(getClass().getResource("/org/financer/client/javafx/main/style/footer.style.css").toExternalForm());
        scene.getStylesheets().add(getClass().getResource("/org/financer/client/javafx/main/style/header.style.css").toExternalForm());
        scene.getStylesheets().add(getClass().getResource("/org/financer/client/javafx/main/style/main.style.css").toExternalForm());
        scene.getStylesheets().add(getClass().getResource("/org/financer/client/javafx/main/style/transactions.style.css").toExternalForm());
        scene.getStylesheets().add(getClass().getResource("/org/financer/client/javafx/main/style/statistics.style.css").toExternalForm());
        scene.getStylesheets().add(getClass().getResource("/org/financer/client/javafx/main/style/overview.style.css").toExternalForm());
        scene.getStylesheets().add(getClass().getResource("/org/financer/client/javafx/main/style/profile.style.css").toExternalForm());
        primaryStage.getIcons().add(new Image(FinancerApplication.class.getResourceAsStream("/images/icons/financer-icon.png")));

        Font.loadFont(getClass().getResource("/fonts/Roboto-Regular.ttf").toExternalForm(), 12);
        Font.loadFont(getClass().getResource("/fonts/Roboto-Medium.ttf").toExternalForm(), 12);

        primaryStage.setTitle("Financer");
        primaryStage.setScene(scene);
        primaryStage.show();
    }
}

File: java/org.financer.client.javafx/src/main/kotlin/org/financer/client/javafx/main/FinancerApplication.kt
package org.financer.client.javafx.main

import javafx.application.Application
import javafx.application.Platform
import javafx.application.Preloader.ProgressNotification
import javafx.beans.property.BooleanProperty
import javafx.beans.property.SimpleBooleanProperty
import javafx.beans.value.ObservableValue
import javafx.fxml.FXMLLoader
import javafx.scene.Parent
import javafx.scene.Scene
import javafx.scene.image.Image
import javafx.scene.paint.Color
import javafx.scene.text.Font
import javafx.stage.Stage
import javafx.stage.StageStyle
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.financer.client.domain.api.RestApi
import org.financer.client.domain.api.RestApiImpl
import org.financer.client.domain.model.category.CategoryRoot
import org.financer.client.domain.model.user.User
import org.financer.client.format.I18N
import org.financer.client.javafx.local.LocalStorageImpl
import org.financer.client.local.LocalStorage
import org.financer.shared.domain.model.value.objects.SettingPair
import java.io.IOException
import java.io.Serializable
import java.util.*
import java.util.logging.Level
import java.util.logging.Logger

class FinancerApplication : Application() {
    private val restApi: RestApi = RestApiImpl()
    private val localStorage: LocalStorage = LocalStorageImpl()
    private val logger = Logger.getLogger("FinancerApplication")
    private val ready: BooleanProperty = SimpleBooleanProperty(false)
    override fun init() {
        val user = LocalStorageImpl.getInstance().readObject<User>("user")
        if (user != null) {
            notifyPreloader(ProgressNotification(0.0))
            CoroutineScope(Dispatchers.IO).launch {
                val categories = restApi.getUsersCategories(user.getId())
                val variableTransactions = restApi.getUsersVariableTransactions(user.getId(), 0)
                val fixedTransactions = restApi.getUsersFixedTransactions(user.getId())
                localStorage.writeObject("categories", CategoryRoot(categories.toSet()))
            }
        }
    }

    @Throws(IOException::class)
    override fun start(primaryStage: Stage) {
        logger.log(Level.INFO, "Financer Application has started.")
        val user = LocalStorageImpl.getInstance().readObject<User>("user")
        if (user == null) {
            initLoginStage()
        } else {
            I18N.setLocalStorage(LocalStorageImpl.getInstance())
            ready.addListener { observableValue: ObservableValue<out Boolean>?, oldValue: Boolean?, newValue: Boolean ->
                if (java.lang.Boolean.TRUE == newValue) {
                    Platform.runLater {
                        try {
                            initMainStage()
                        } catch (e: IOException) {
                            logger.log(Level.SEVERE, e.message, e)
                        }
                    }
                }
            }
        }
    }

    @Throws(IOException::class)
    private fun initLoginStage() {
        // setting up language
        val resourceBundle = ResourceBundle.getBundle("Financer", Locale.ENGLISH)

        // loading FXML file
        val root = FXMLLoader.load<Parent>(javaClass.getResource("/org/financer/client/javafx/main/views/login.fxml"), resourceBundle)
        val primaryStage = Stage()
        val scene = Scene(root, 800.0, 500.0)
        scene.stylesheets.add(javaClass.getResource("/org/financer/client/javafx/main/style/colors.style.css").toExternalForm())
        scene.stylesheets.add(javaClass.getResource("/org/financer/client/javafx/main/style/form.style.css").toExternalForm())
        scene.stylesheets.add(javaClass.getResource("/org/financer/client/javafx/main/style/login.style.css").toExternalForm())
        scene.fill = Color.TRANSPARENT
        primaryStage.icons.add(Image(FinancerApplication::class.java.getResourceAsStream("/images/icons/financer-icon.png")))
        Font.loadFont(javaClass.getResource("/fonts/Roboto-Regular.ttf").toExternalForm(), 12.0)
        Font.loadFont(javaClass.getResource("/fonts/Roboto-Medium.ttf").toExternalForm(), 12.0)
        primaryStage.initStyle(StageStyle.TRANSPARENT)
        primaryStage.title = "Financer - Login"
        primaryStage.scene = scene
        primaryStage.show()
    }

    @Throws(IOException::class)
    private fun initMainStage() {
        val resourceBundle = ResourceBundle.getBundle("Financer",
                (LocalStorageImpl.getInstance().readObject<Serializable>("user") as User?)!!.getValueOrDefault<Any>(SettingPair.Property.LANGUAGE) as Locale)
        val root = FXMLLoader.load<Parent>(javaClass.getResource("views/financer.fxml"), resourceBundle)
        val primaryStage = Stage()
        val scene = Scene(root, 1300.0, 700.0)
        scene.stylesheets.add(javaClass.getResource("/org/financer/client/javafx/main/style/colors.style.css").toExternalForm())
        scene.stylesheets.add(javaClass.getResource("/org/financer/client/javafx/main/style/form.style.css").toExternalForm())
        scene.stylesheets.add(javaClass.getResource("/org/financer/client/javafx/main/style/navbar.style.css").toExternalForm())
        scene.stylesheets.add(javaClass.getResource("/org/financer/client/javafx/main/style/footer.style.css").toExternalForm())
        scene.stylesheets.add(javaClass.getResource("/org/financer/client/javafx/main/style/header.style.css").toExternalForm())
        scene.stylesheets.add(javaClass.getResource("/org/financer/client/javafx/main/style/main.style.css").toExternalForm())
        scene.stylesheets.add(javaClass.getResource("/org/financer/client/javafx/main/style/transactions.style.css").toExternalForm())
        scene.stylesheets.add(javaClass.getResource("/org/financer/client/javafx/main/style/statistics.style.css").toExternalForm())
        scene.stylesheets.add(javaClass.getResource("/org/financer/client/javafx/main/style/overview.style.css").toExternalForm())
        scene.stylesheets.add(javaClass.getResource("/org/financer/client/javafx/main/style/profile.style.css").toExternalForm())
        primaryStage.icons.add(Image(FinancerApplication::class.java.getResourceAsStream("/images/icons/financer-icon.png")))
        Font.loadFont(javaClass.getResource("/fonts/Roboto-Regular.ttf").toExternalForm(), 12.0)
        Font.loadFont(javaClass.getResource("/fonts/Roboto-Medium.ttf").toExternalForm(), 12.0)
        primaryStage.title = "Financer"
        primaryStage.scene = scene
        primaryStage.show()
    }
}